core/src/main/java/com/linecorp/armeria/common/stream/DeferredStreamMessage.java

```diff
@@ -317,6 +333,36 @@ public final void abort(Throwable cause) {
         }
     }
 
+    @Override
+    public CompletableFuture<List<T>> collect(EventExecutor executor, SubscriptionOption... options) {
+        requireNonNull(executor, "executor");
+        requireNonNull(options, "options");
+
+        if (downstreamSubscription != null) {
```

Comment:
Shouldn't we CAS with some no-op subscription?

Reply:
I thought early return with null checking is enough here because `DeferredStreamMessage` is not a primitive implementation.
The race condition would rarely happen and primitive `StreamMessage`s such as `FixedStreamMesage`, `DefaultStreamMessage` and `PublisherBasedStreamMessage` would check it finally.

Reply:
> The race condition would rarely happen and primitive StreamMessages such as FixedStreamMesage, DefaultStreamMessage and PublisherBasedStreamMessage would check it finally.

If the late subscribing is handled by the stream messages, how about removing `downstreamSubscriptionUpdater` also?
I think it's weird that we only check just one way for the race but let the other way just pass through. 🤔

Reply:
>  how about removing downstreamSubscriptionUpdater also?

That sounds good. I also had a similar idea. Let me delegate the validation on a single subscription to the upstream.
I guess it would be better for performance with a normal code path.